Use `custom_logo` theme-mod instead of `site_logo` setting for the site-logo block

aristath opened the pull request:
## Description

In the previous iteration we were adding a `site_logo` setting, and then we were trying to sync the setting with the pre-existing `custom_logo` theme-mod. This PR removes the `site_logo` setting, and uses the `custom_logo` theme-mod to get and set the site-logo.
To do that, `theme_mods.custom_logo` and `stylesheet` were added to the Settings REST API, and then the block's edit JS was tweaked accordingly.

## How has this been tested?

Tested by adding a site-logo block, setting and changing the logo.
Then went in the database and confirmed that values were properly formatted in all cases.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aristath committed:
Use custom_logo theme-mod instead of site_logo setting

aristath committed:
fix issue when theme-mod is unset

aristath commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -71,56 +71,83 @@ function register_block_core_site_logo() {
-		delete_option( 'site_logo' );
-	} else {
-		update_option( 'site_logo', $custom_logo );
+function gutenberg_rest_pre_get_setting_filter_custom_logo( $result, $name ) {
```
This function was necessary because the REST API doesn't return the value if it has more items than the ones that were defined when registering the setting, and we can't know what theme-mods will be available or how they will be formatted.
We only need the `custom_logo` setting and this is safer than exposing all theme-mods.

aristath commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -71,56 +71,83 @@ function register_block_core_site_logo() {
-			'description'  => __( 'Site logo.' ),
-		)
-	);
+function gutenberg_rest_pre_set_setting_filter_theme_mods( $result, $name, $value ) {
```
This function is used so that the theme-mods we set get merged with the pre-existing ones instead of replacing them. Without this, after saving the logo we'd only have the `custom_logo` theme-mod in the database and everything else would be getting trashed.

aristath commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -71,56 +71,83 @@ function register_block_core_site_logo() {
- * Syncs the site logo with the theme modified logo.
+ * Expose the "stylesheet" setting in the REST API.
+ */
+register_setting(
```
The `stylesheet` setting is never properly registered in core, so we need to register it here and expose it to the REST API.

aristath commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -71,56 +71,83 @@ function register_block_core_site_logo() {
-	$site_logo = get_option( 'site_logo' );
-	return false === $site_logo ? $custom_logo : $site_logo;
-}
+register_setting(
```
Registers the `custom_logo` theme-mod as a setting. The value is then filtered in functions below, to get and set the theme-mod.

gziolo commented:
Nice, thank you for opening this PR. It looks like an interesting alternative as we always operate on a single setting specific to the current theme 👍🏻 

One thing I'm contemplating now is whether we could flatten this new setting included in REST API. We add handlers for set and get so in theory we could expose it as a top-level setting and handle the complex logic in PHP. What do you think?

We could also move updates to REST API endpoint to `lib/` folder so they aren't copied to WordPress core together with `index.php` for the block.

aristath committed:
expose custom_logo as a top-level item.

aristath commented:
> One thing I'm contemplating now is whether we could flatten this new setting included in REST API. We add handlers for set and get so in theory we could expose it as a top-level setting and handle the complex logic in PHP. What do you think?

Done. It's a lot cleaner now :+1:

aristath committed:
Move REST-API tweaks from the block to lib

aristath commented:
> We could also move updates to REST API endpoint to lib/ folder so they aren't copied to WordPress core together with index.php for the block.

Good point, there's no reason for these to be in the block itself :+1: Done.

creativecoder commented:
I think the main disadvantage to this approach is that the `custom_logo` setting is stored in a theme specific option (`theme_mods_{theme}`). This means that when you switch themes, the logo disappears from the site logo block. I believe that's the main reason the block was created to use a dedicated option.

aristath commented:
> I think the main disadvantage to this approach is that the `custom_logo` setting is stored in a theme specific option (`theme_mods_{theme}`). This means that when you switch themes, the logo disappears from the site logo block. I believe that's the main reason the block was created to use a dedicated option.

Yes. That is the way it was always working in WordPress, logos are theme-specific because themes can (and do) change the way logos are displayed, therefore the logo used for one theme is not the same that should be used for all themes :+1:

creativecoder commented:
> That is the way it was always working in WordPress, logos are theme-specific because themes can (and do) change the way logos are displayed, therefore the logo used for one theme is not the same that should be used for all themes 👍

I see the logic of that when the display of the logo is dependent on the theme's templates, as it has been with the `custom-logo` functionality.

But with a Site Logo block that's placed within the content or in an editor based template, the logo disappearing from the block on theme switch seems very disruptive to me. I think we have an opportunity to improve the experience when switching themes by making the block setting independent from the theme.

TimothyBJacobs commented:
Are we just reusing the settings endpoint because it is an existing endpoint we can attach on to? Am I understanding correctly that it will now always be stored in a theme mod, not in an option?

aristath commented:
We're reusing the settings endpoint because **theme-mods are settings** (`theme_mod_STYLESHEET` setting)

TimothyBJacobs commented:
Ok, I'm a pretty firm -1 on this then. We shouldn't be using the settings endpoint just because it allows us to use an endpoint. The fact that theme mods are eventually stored as settings isn't really relevant since we aren't exposing or utilizing the theme mod option.

The themes endpoint should instead allow for updating theme mods that have been registered. For instance thru a `register_theme_mod` which would be an analogue for `register_setting` or `register_theme_feature`.

aristath committed:
revert https://github.com/WordPress/gutenberg/pull/31994/commits/50b57c50b650d09e95c646fbb36343a50889a9c5

aristath commented:
> The fact that theme mods are eventually stored as settings isn't really relevant since we aren't exposing or utilizing the theme mod option.

Thank you for the feedback @TimothyBJacobs. That's what we were initially doing, but then moved the `custom_logo` to the top level instead of having it nested inside the `theme_mod_*` option. I reverted that commit, so the settings endpoint now has the theme-mods option instead - as it should.
I believe that addresses the concerns expressed for the REST API?

As for the alternatives, what we previously had was exposing a new `site_logo` option in the REST API, which was then used to sync the value to the theme-mod. It worked, but exposing a new option just for the sake of keeping the API "pure" and in the background sync options left and right is not a good implementation either.
On the other hand, adding theme-mods to the themes endpoint just because they have "themes" in their name is also sub-optimal since the themes endpoint is used to query themes and not the settings that themes save.
Using the settings endpoint makes more sense now that we expose the theme-mod option...
Perhaps in the future we could expose a new endpoint for theme-mods? but that's probably not something that can (or should) be done right now.

youknowriad reviewed (approved):
Ok looks like the feedback is addressed and we need this for RC to ease integration with Core.

creativecoder commented:
It seems to me that in the rush to integrate with Core (which I understand!), we're sacrificing the user experience of this block.

Here's what happens when you switch themes. I understand this has been normalized by how custom-logo works now in the Customizer, but in the context of blocks and the editor, it seems quite broken to me.
 
![logo-disappears-when-switching-themes](https://user-images.githubusercontent.com/1699996/119006288-dad9d600-b955-11eb-85ca-a0963cd298aa.gif)

---

(Also the block is broken on trunk right now if you don't have a logo set, but that's a relatively minor fix)

creativecoder commented:
I created a separate issue so we can discuss this more: https://github.com/WordPress/gutenberg/issues/32065
